### PR TITLE
wayland presentation fixes (again)

### LINKED ISSUE #11022: Leak of last allocated struct wl_presentation_feedback on exit
### Important Information

Provide following Information:
- Version: 0.35.0 / f9d0b0c08a35e67976a8c71ac16a832781c2313a
- Linux Distribution and Version: ArchLinux latest
- Source of the mpv binary: Self built
- If known which version of mpv introduced the problem: -
- Window Manager and version: sway 1.7, wlroots 0.15.1
- GPU driver and version: Mesa Intel(R) UHD Graphics 600 (GLK 2) /  Mesa 22.3.1
- Possible screenshot or video of visual glitches: N/A

### Reproduction steps

Compile mpv with ASan. Run it with vo=gpu under wayland.
Exit mpv.

### Expected behavior

No ASan warnings are reported.

### Actual behavior

An ASan leak is reported. (See log)

### Log file

Note: this logfile contains logs of all allocated `struct wl_proxy` instaces from `proxy_create()` in `wayland-client.c`.
As can be seen the leaked object is the second-to-last wl_presentation_feedback that was allocated in total and the last that was allocated before mpv exits.

It seems for this wl_presentation_feedback the `feedback_presented` callback is never called and it is also not freed in `uninit()` of `wayland_common.c` because another one is allocated before uninit() is called.

```
LSAN_OPTIONS=report_objects=1 ./mpv --no-config  $SOME_FILE
 (+) Video --vid=1 (*) (h264 1280x720 29.998fps)
 (+) Audio --aid=1 (*) (aac 2ch 48000Hz)
proxy_create wl_registry:0 0x60700001e240
proxy_create wl_callback:0 0x60700001e320
proxy_create wl_shm:1 0x60700001e390
proxy_create zwp_linux_dmabuf_v1:4 0x60700001e400
proxy_create zwp_linux_dmabuf_feedback_v1:4 0x60700001e470
proxy_create wl_compositor:4 0x60700001e4e0
proxy_create wl_surface:4 0x60700001e550
proxy_create wl_surface:4 0x60700001e5c0
proxy_create wl_region:4 0x60700001e630
proxy_create wl_surface:4 0x60700001e6a0
proxy_create wl_subcompositor:1 0x60700001e710
proxy_create wl_data_device_manager:3 0x60700001e780
proxy_create zwp_idle_inhibit_manager_v1:1 0x60700001e7f0
proxy_create xdg_wm_base:2 0x60700001e860
proxy_create zxdg_decoration_manager_v1:1 0x60700001e8d0
proxy_create wp_presentation:1 0x60700001e940
proxy_create wp_viewporter:1 0x60700001e9b0
proxy_create wl_seat:1 0x60700001ea20
proxy_create wl_output:4 0x60700001ea90
proxy_create xdg_surface:2 0x60700001eb00
proxy_create xdg_toplevel:2 0x60700001eb70
proxy_create wl_subsurface:1 0x60700001ebe0
proxy_create wp_viewport:1 0x60700001ec50
proxy_create wp_viewport:1 0x60700001ecc0
proxy_create wl_data_device:3 0x60700001ed30
proxy_create zxdg_toplevel_decoration_v1:1 0x60700001eda0
proxy_create wl_callback:4 0x60700001ee10
proxy_create wl_callback:0 0x60700001eef0
proxy_create wl_pointer:1 0x60700001ef60
proxy_create wl_keyboard:1 0x60700001efd0
proxy_create wl_registry:0 0x60700001f2e0
proxy_create wl_callback:0 0x60700001f3c0
proxy_create zwp_linux_dmabuf_v1:4 0x60700001f430
proxy_create zwp_linux_dmabuf_feedback_v1:4 0x60700001f4a0
proxy_create wl_callback:0 0x60700001f580
AO: [pipewire] 48000Hz stereo 2ch floatp
VO: [gpu] 1280x720 yuv420p
proxy_create zwp_linux_dmabuf_feedback_v1:4 0x60700005e820
proxy_create wl_callback:0 0x60700005e900
proxy_create wl_region:4 0x60700005e970
proxy_create zwp_linux_buffer_params_v1:4 0x60700006d0e0
proxy_create wl_buffer:4 0x60700006d150
proxy_create wl_callback:0 0x60700006d1c0
proxy_create wl_callback:4 0x60700006d230
proxy_create wp_presentation_feedback:1 0x60700006d2a0
proxy_create wl_region:4 0x60700006d310
proxy_create zwp_linux_buffer_params_v1:4 0x60700006d3f0
proxy_create wl_buffer:4 0x60700006d460
proxy_create wl_callback:0 0x60700006d4d0
proxy_create wl_callback:4 0x60700006d540
proxy_create wp_presentation_feedback:1 0x60700006d5b0
proxy_create zwp_linux_buffer_params_v1:4 0x60700006d690
proxy_create wl_buffer:4 0x60700006d700
proxy_create wl_callback:0 0x60700006d770
proxy_create wl_callback:4 0x60700006d7e0
proxy_create wp_presentation_feedback:1 0x60700006d850
proxy_create zwp_idle_inhibitor_v1:1 0x60700006d8c0
proxy_create wl_callback:0 0x60700006d9a0
proxy_create wl_callback:4 0x60700006da10
proxy_create wp_presentation_feedback:1 0x60700006da80
proxy_create wl_callback:0 0x60700006db60
proxy_create wl_callback:4 0x60700006dbd0
proxy_create wp_presentation_feedback:1 0x60700006dc40
proxy_create wl_callback:0 0x60700006dd20
proxy_create wl_callback:4 0x60700006dd90
proxy_create wp_presentation_feedback:1 0x60700006de00
proxy_create wl_callback:0 0x60700006dee0


Exiting... (Quit)
proxy_create wl_callback:4 0x607000070020
proxy_create wp_presentation_feedback:1 0x607000070090

=================================================================
==25940==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 80 byte(s) in 1 object(s) allocated from:
    #0 0x7f405bcbf411 in __interceptor_calloc /usr/src/debug/gcc/libsanitizer/asan/asan_malloc_linux.cpp:77
    #1 0x7f4059587baa in zalloc ../wayland-1.21.0/src/wayland-private.h:234
    #2 0x7f4059587baa in proxy_create ../wayland-1.21.0/src/wayland-client.c:422

Objects leaked above:
0x60700006de00 (80 bytes)  // Note: This is a `struct wl_proxy`.

SUMMARY: AddressSanitizer: 80 byte(s) leaked in 1 allocation(s).
```

### PR DESCRIPTION
**Pull request by Dudemanguy**
Fixes #11022 but it's better this time.

First commit is just extra. The actual fix is the second one which basically just dispatches wayland events and then does another roundtrip before doing uninit. A `wl->feedback == fback` check is added now since these two are not necessarily the same pointer.

### CODE REVIEW

File: video/out/wayland_common.c [hunk 1]
```diff
@@ -1022,17 +1022,14 @@ static void feedback_presented(void *data, struct wp_presentation_feedback *fbac
...
     struct vo_wayland_state *wl = data;
 
-    // NULL is needed to prevent a dangling pointer since presentation_feedback
-    // is created in the frame_callback and not in any of the actual presentation
-    // events.
+    // NULL is needed to prevent a double free during uninit since wl->feedback will
+    // be the same as fback.
     if (fback) {
+        if (wl->feedback == fback)
+            wl->feedback = NULL;
```
Comment (t-8ch) [inline video/out/wayland_common.c:1029]: Is wl->feedback still useful at all?
Should the new logic not take care of all feedbacks?

Comment (Dudemanguy) [reply video/out/wayland_common.c:1029]: There's always one last feedback that's created and we won't get a discard or presented event. This essentially keeps track of that. The frame callback is the same way.

Comment (t-8ch) [reply video/out/wayland_common.c:1029]: Fair enough. Although it seems weird that this behavior would be intentional.

Comment (Dudemanguy) [reply video/out/wayland_common.c:1029]: I mean it's unavoidable. After every frame is drawn, we have to setup a new presentation feedback listener which will fire after the next draw occurs. If the user quits in the middle of this, well that next draw never actually occurs so you won't get a discarded or presented event from that. You have to just free it along with all the other `wl_proxy`'s.

Comment (t-8ch) [reply video/out/wayland_common.c:1029]: But that last listener is still registered to the surface, and destroying the surface should trigger a discard event.

Indeed adding a `wl_display_roundtrip(); wayland_dispatch_events();` after `wl_surface_destroy()` in `uninit()` seems to work.

I think the order has to be `roundtrip()` and then `dispatch()`, instead of the other way round as the PR does.
`roundtrip()` should make sure all events are delivered and `dispatch()` will handle them on mpv side.
Otherwise we may loose events (which I can reproduce)

Comment (t-8ch) [reply video/out/wayland_common.c:1029]: Then all of the `NULL` checks and special handling in the feedback listeners seem to be unnecessary, too.

Comment (t-8ch) [reply video/out/wayland_common.c:1029]: Maybe even better yet:
```c
if (roundtrip())
    dispatch()
```
This should make sure that we are never waiting in the dispatcher.

Comment (Dudemanguy) [reply video/out/wayland_common.c:1029]: This leaks for me if I move the surface to another desktop, go to that desktop, and then quit (sway). My PR leaks in that case as well though.

Comment (t-8ch) [reply video/out/wayland_common.c:1029]: Hm, it doesn't leak for me. With "desktop" do you mean virtual workspace or physical screen?

Could you push your version somewhere? Mine is here: https://github.com/t-8ch/mpv/tree/wayland-presentation-test

Comment (Dudemanguy) [reply video/out/wayland_common.c:1029]: Not really different than yours. Just this.

```
diff --git a/video/out/wayland_common.c b/video/out/wayland_common.c
index b1a3a6a2ff..bcac963285 100644
--- a/video/out/wayland_common.c
+++ b/video/out/wayland_common.c
@@ -1024,11 +1024,8 @@ static void feedback_presented(void *data, struct wp_presentation_feedback *fbac
 
     // NULL is needed to prevent a double free during uninit since wl->feedback will
     // be the same as fback.
-    if (fback) {
-        if (wl->feedback == fback)
-            wl->feedback = NULL;
+    if (fback)
         wp_presentation_feedback_destroy(fback);
-    }
 
     wl->refresh_interval = (int64_t)refresh_nsec / 1000;
 
@@ -1052,11 +1049,8 @@ static void feedback_discarded(void *data, struct wp_presentation_feedback *fbac
     struct vo_wayland_state *wl = data;
 
     // Same logic in feedback_presented applies here.
-    if (fback) {
-        if (wl->feedback == fback)
-            wl->feedback = NULL;
+    if (fback)
         wp_presentation_feedback_destroy(fback);
-    }
 }
 
 static const struct wp_presentation_feedback_listener feedback_listener = {
@@ -2153,9 +2147,6 @@ void vo_wayland_uninit(struct vo *vo)
     if (!wl)
         return;
 
-    wayland_dispatch_events(wl, 1, 10);
-    wl_display_roundtrip(wl->display);
-
     mp_input_put_key(wl->vo->input_ctx, MP_INPUT_RELEASE_ALL);
 
     if (wl->compositor)
@@ -2190,8 +2181,8 @@ void vo_wayland_uninit(struct vo *vo)
     if (wl->dnd_offer)
         wl_data_offer_destroy(wl->dnd_offer);
 
-    if (wl->feedback)
-        wp_presentation_feedback_destroy(wl->feedback);
+//    if (wl->feedback)
+//        wp_presentation_feedback_destroy(wl->feedback);
 
     if (wl->frame_callback)
         wl_callback_destroy(wl->frame_callback);
@@ -2251,6 +2242,9 @@ void vo_wayland_uninit(struct vo *vo)
     if (wl->video_subsurface)
         wl_subsurface_destroy(wl->video_subsurface);
 
+    if (wl_display_roundtrip(wl->display))
+        wayland_dispatch_events(wl, 1, 10);
+
     if (wl->wm_base)
         xdg_wm_base_destroy(wl->wm_base);
 
```

I meant virtual workspace my bad. If I change `10` to `100`, it doesn't leak.

Comment (t-8ch) [reply video/out/wayland_common.c:1029]: > If I change 10 to 100, it doesn't leak.

This feels a bit hacky. If I understand correctly the poll should return instantly anyways because we already forced all events through and they should be ready to be read.

It should even work with changing the 10 to 0, I think. It seems we are missing something.

Comment (Dudemanguy) [reply video/out/wayland_common.c:1029]: >This feels a bit hacky.

Yeah agreed. I think this is the right track though. If we can come up with something that isn't dependent on some timeout that always process the events, that should solve it.

Comment (t-8ch) [reply video/out/wayland_common.c:1029]: Putting in a `sleep(1)` before the dispatching makes the leak go away. So the discard events do all get sent.

Comment (t-8ch) [reply video/out/wayland_common.c:1029]: Also on mutter/gnome it works for me reliable even with timeout=0. Could you test that too?
Maybe a bug in wlroots?

Comment (Dudemanguy) [reply video/out/wayland_common.c:1029]: Indeed, it seems to always work in mutter.

Comment (Dudemanguy) [reply video/out/wayland_common.c]: Some unfortunate news, this method doesn't seem to work in weston at all. Sway will eventually deliver the event if you wait long enough (in this case, the second-to-last feedback creation) so in theory a better block would solve it. In weston this doesn't appear to happen at all. It never sends a discard.

Comment (t-8ch) [reply video/out/wayland_common.c]: But that would mean on weston presentation-time will leak unbounded amounts of memory anyways.
This seems like a fairly serious bug. Cleaning it up properly at `uninit()` is a very weak bandaid.

Comment (Dudemanguy) [reply video/out/wayland_common.c]: Oh sorry, I phrased that badly. I meant the last discard isn't cleaned (AKA it will always leak the same amount of memory anyway). However, I think this one is actually a weston bug. The code looks like it attempts to send a discard when the surface is destroyed but we never get it for some reason. With sway/wlroots, I'm not sure exactly what's going on. It gives us the discard immediately but for some unspecified period of time later, it gives us the second to last presented event (which occurs after the discarded one). Ironically, mutter is the one that works as I would expect. Go figure.

File: video/out/wayland_common.c [hunk 2]
```diff
@@ -2155,6 +2146,21 @@ void vo_wayland_uninit(struct vo *vo)
...
+    if (wl->xdg_toplevel)
+        xdg_toplevel_destroy(wl->xdg_toplevel);
+
+    if (wl->xdg_surface)
+        xdg_surface_destroy(wl->xdg_surface);
+
+    if (wl->surface)
+        wl_surface_destroy(wl->surface);
+
+    // Proccess any final events.
```
Comment (t-8ch) [inline video/out/wayland_common.c:2160]: "Process"

File: video/out/wayland_common.c (hunk 1 above)
Comment (t-8ch) [reply video/out/wayland_common.c]: 3c62bee works nicely for me on sway.

Comment (Dudemanguy) [reply video/out/wayland_common.c]: Not for me. Unless I bump up the timeout, but I'm trying to avoid that (no luck yet).

Comment (t-8ch) [reply video/out/wayland_common.c]: FYI the behavior of weston is definitively a bug. It leaks the surface in the fade animation somewhere.

Comment (Dudemanguy) [reply video/out/wayland_common.c]: Nice. I'm also pretty sure that sway/wlroots is bugged too here. When destroying the surface (aka quitting mpv), you get logs like this (present/discard added by me).

```
00:00:07.771 [DEBUG] [sway/desktop/idle_inhibit_v1.c:22] Sway idle inhibitor destroyed
00:00:07.771 [DEBUG] [wlr] [types/wlr_idle.c:183] Enabling idle timers for all seats
00:00:07.771 [DEBUG] [sway/desktop/idle_inhibit_v1.c:22] Sway idle inhibitor destroyed
00:00:07.771 [DEBUG] [sway/tree/arrange.c:263] Usable area for ws: 1920x1059@0,0
00:00:07.771 [DEBUG] [sway/tree/arrange.c:293] Arranging workspace '1' at 0.000000, 0.000000
00:00:07.771 [DEBUG] [sway/tree/arrange.c:77] Arranging 0x7fffa441fa50 horizontally
00:00:07.771 [DEBUG] [sway/desktop/transaction.c:398] Transaction 0x55cd5a8564e0 committing with 5 instructions
00:00:07.771 [DEBUG] [sway/desktop/transaction.c:294] Applying transaction 0x55cd5a8564e0
00:00:07.771 [INFO] [wlr] [types/wlr_presentation_time.c:244] Send discard event.
00:00:07.779 [INFO] [wlr] [types/wlr_presentation_time.c:300] Send present event.
```
That final present event (which corresponds to the second-to-last feedback mpv creates) gets sent *after* sway commits its atomic transaction for surface destruction. That's why adding a slight timeout works. When you screw around with the surface, the time it takes to send the final present event can increase. When a surface is destroyed, it should send any presented events at the same time as discarded imo.

tl;dr: I think this patch is right now but there's a couple of compositor bugs (I'll try to fix wlroots).

Comment (t-8ch) [reply video/out/wayland_common.c]: Do you already have a simple reproduction case?

**Comment by Dudemanguy** [comment]
I went ahead and put a 50 ms timeout for now as a workaround. I can try to work on sending patches for wlroots/weston in the meantime.

**Comment by t-8ch** [comment]
> I went ahead and put a 50 ms timeout for now as a workaround.

Sounds good.

It feels a bit weird to replace one memory leak with another, equivalent one, though :-)

**Comment by t-8ch** [comment]
When you have something for wlroots/sway I would be curious.

**Comment by Dudemanguy** [comment]
Yeah I really doubt that's going to happen. The surface lifetimes are pretty much not ordered at all. I should probably actually be a good citizen and report it though.

**Comment by t-8ch** [comment]
One of the weston example clients does also keep a list of presentation feedbacks around to clean them up manually at shutdown: https://gitlab.freedesktop.org/wayland/weston/-/blob/main/clients/presentation-shm.c#L311

For me the protocol description is unclear about the lifecycle rules.
Maybe it's best to open a ticket for the protocol to ask for clarification. WDYT?

**Comment by Dudemanguy** [comment]
Yeah manually keeping track might be the more "correct" way but we can't handle it like weston's example client does since we don't set the listener in the presentation events itself. It's probably worth asking; I would suspect mpv is one of the very few clients to actually use presentation-time in the first place and existing compositors appear to all have wildly differing behaviors.

**Comment by t-8ch** [comment]
Ticket: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/issues/125
